src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_CAPACITY];
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        if (size == storage.length) {
+            grow();
+        }
+        storage[size++] = value;
+    }
 
+    @Override
+    public void add(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        if ((index < 0) || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index! " + index);
+        }
+        if (size == storage.length) {
+            grow();
+        }
+        System.arraycopy(storage, index, storage, index + 1, size - index);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (list.size() > (storage.length - size)) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            storage[size] = list.get(i);
+            size++;
+        }
```

Comment:
just call add method in the loop, all the necessary logic is there, if add is implemented correctly, so will be this  method

Reply:
wow, this is a really cool punch, nice! thanks!